File: c++/N-Queens_II.cpp
/* Follow up for N-Queens problem.

Now, instead outputting board configurations, return the total number of distinct solutions.

*/
class Solution {
public:
    void findSolution(int n, vector<int> pos, int row, int& res){
        if(pos.size() == n){
            res++;
        }
        for(int i  = 0; i < n; i++){
            bool valid = true;
            for(int j = 0; j < pos.size(); j++){
                if(i == pos[j] || abs(i - pos[j]) == row - j){
                    valid = false;
                    break;
                }
            }
            if(valid){
                pos.push_back(i);
                findSolution(n,pos,row + 1, res);
                pos.pop_back();
            }
        }
    }
    int totalNQueens(int n) {
        int res = 0;
        vector<int> pos;
        findSolution(n,pos,0,res);
        return res;
    }
};

File: python/N-Queens_II.py
"""
Follow up for N-Queens problem.

Now, instead outputting board configurations, return the total number of distinct solutions.



Subscribe to see which companies asked this question
"""
class Solution(object):
    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        def solver(pos, row, n, res):
            if row == n :
                res[0] += 1
            for i in range(n):
                valid = True
                for j in range(len(pos)):
                    if i == pos[j] or abs(i - pos[j]) == row - j:
                        valid = False
                        break
                if valid:
                    pos.append(i)
                    solver(pos, row + 1, n, res)
                    pos.pop()
        res = [0]
        pos = []
        solver(pos, 0, n, res)
        return res[0]